ls1intum/Artemis
# `General`: Add an optional start date to exercises

### Checklist
#### General


#### Server
#### Client

### Motivation and Context
#5803

### Description

Resolves #5803

This PR adds a "Start Date" to Exercises except Quizzes. The start date is always after the release date.
After release date, the task is visible but the students cannot start until the start date is reached.
Nevertheless, the students are still able to use communications during the period between release and start date.
Tutors and instructors can always start the exercise.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Exercise (all but quiz exercise) with enabled communication features enabled

1. Log in to Artemis
2. Create an exercise in a course with enabled communication feature
3. Set Release date and Start date
4. Check that communication is possible after release date & starting is not possible before start date.

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| programming-exercise-lifecycle.component.ts | 78% | 93% |


### Screenshots
#### After Release, Before Startdate
![Bildschirm­foto 2022-11-21 um 12 32 58](https://user-images.githubusercontent.com/38322605/203042405-2f379786-f431-4dcc-b3d2-5de4020794fc.png)

#### After Startdate
![Bildschirm­foto 2022-11-21 um 12 39 41](https://user-images.githubusercontent.com/38322605/203042452-763a1f56-de40-4f6e-a957-120e82975917.png)

#### Instructor view
![Bildschirm­foto 2022-11-21 um 12 40 57](https://user-images.githubusercontent.com/38322605/203042590-b1eddbc7-1eeb-4751-86a0-f78375b6d4c9.png)
